flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java

```diff
@@ -381,12 +333,28 @@ private ForwardingValveOutputHandler(
 			this.streamStatusMaintainer = checkNotNull(streamStatusMaintainer);
 
 			this.inputWatermarkGauge = inputWatermarkGauge;
+			this.numRecordsIn = checkNotNull(numRecordsIn);
 
 			this.inputIndex = inputIndex;
 		}
 
 		@Override
-		public void handleWatermark(Watermark watermark) throws Exception {
+		@SuppressWarnings("unchecked")
+		public void emitRecord(StreamRecord record) throws Exception {
+			synchronized (lock) {
+				numRecordsIn.inc();
+				if (inputIndex == 0) {
```

Comment:
this might add some extra overhead compared to the previous code, as probably JIT won't be able to correctly handle this and branch prediction will not work (as far as I know, most/all optimisations are per class, not per class instance, so if there are two instances of the same class, where one has `inputIndex == 0` and other has `inputIndex == 1`, they will collide with one another).

Maybe it would be better to have this class abstract, with abstract method for `processElement`, which would be implemented concretely in `FirstStreamTaskNetworkOutput` and `SecondStreamTaskNetworkOutput` as
```
					operator.setKeyContextElement1(record);
					operator.processElement1(record);
```
and 
```
					operator.setKeyContextElement2(record);
					operator.processElement2(record);
```
respectively? We could leave watermarks/statuses/latency markers as they are, since they are not performance critical.

Normally I would suggest to relay on our benchmarks, but unfortunately our benchmarks for this selectable processor are very unstable :(

Reply:
Yes, I think you might be right for this performance concern. If we have two separate subclasses to implement abstract `emitRecords`, it could work.
Another possible option is to use the consumer function in my first version, then we could implement this function in `StreamTwoInputSelectableProcessor` and pass it into output constructor. This way seems a bit simple because it only relies on one class `StreamTaskNetworkOutput`, otherwise we need three classes (one abstract class and two subclasses). WDYT?

Reply:
I think this is fine. Drawback is that stacktraces/debugging will be a little bit more difficult, but we can try this out.

Reply:
Yes, my only concern for this way is also for potential debugging issue.